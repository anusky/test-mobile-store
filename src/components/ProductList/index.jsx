import { string, arrayOf, shape } from "prop-types";
import { useEffect, useMemo, useRef, useState } from "react";
import { Link } from "react-router-dom";

const THRESHOLD_ELEMENTS = 12;
const ProductList = ({ list }) => {
  const originList = useRef(list);
  const [currentList, setCurrentlist] = useState(
    list.slice(0, THRESHOLD_ELEMENTS)
  );
  const currentListLength = useRef(THRESHOLD_ELEMENTS);
  const firstUpdate = useRef(true);
  const bottomElement = useRef(null);

  useEffect(() => {
    const callback = (entries, _observer) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting && !firstUpdate.current) {
          const newLength = currentListLength.current + THRESHOLD_ELEMENTS;
          currentListLength.current =
            newLength > originList.current.length
              ? originList.current.length
              : newLength;
          setCurrentlist(
            originList.current.slice(0, currentListLength.current)
          );
        } else {
          firstUpdate.current = false;
        }
      });
      /**
       * Stop obersvers when we have rendered all tuuhe list elements
       */
      currentListLength.current === originList.current.length &&
        _observer.unobserve(bottomElement.current);
    };
    const options = {
      root: null,
      rootMargin: "0px",
      threshold: 1.0,
    };

    const observer = new IntersectionObserver(callback, options);
    /**
     * Prevent from creating an observer when there are not enough elements to list
     */
    if (list.length > THRESHOLD_ELEMENTS) {
      observer.observe(bottomElement.current);
    } else {
      observer.unobserve(bottomElement.current);
    }
  }, [originList.current]);

  useEffect(() => {
    originList.current = list;
    currentListLength.current =
      list.length > THRESHOLD_ELEMENTS ? THRESHOLD_ELEMENTS : list.length;
    setCurrentlist(list.slice(0, currentListLength.current));
  }, [list.length]);

  const itemList = useMemo(() => {
    return currentList.map((product, index) => {
      return (
        <Link
          key={index}
          to={`/${product.id}`}
          className="grid gap-y-4 gap-x-4 justify-center hover:bg-slate-50 rounded-lg py-4 px-3"
        >
          <div className="object-cover">
            <img
              src={product.imgUrl}
              alt={`${product.brand}, ${product.model}`}
            />
          </div>
          <div className="grid">
            <p>
              {product.brand}, {product.model}
            </p>
            <p className="font-bold">{product.price}â‚¬</p>
          </div>
        </Link>
      );
    });
  }, [currentList]);
  return (
    <div
      className="grid grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-y-8 justify-center"
      data-testid="productlist-component"
    >
      {itemList}
      <div ref={bottomElement}></div>
    </div>
  );
};

export default ProductList;
ProductList.propTypes = {
  list: arrayOf(
    shape({
      id: string,
      brand: string,
      model: string,
      price: string,
      imgUrl: string,
    })
  ),
};
