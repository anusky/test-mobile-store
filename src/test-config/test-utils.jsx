import "./__mocks__/intersectionObserverMock";

import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import PropTypes from "prop-types";

import productInfo from "./__fixtures__/product.json";
import productList from "./__fixtures__/productList.json";
import cart from "./__fixtures__/cart.json";
import { ProductContextProvider } from "src/context/productContext";
import { PAGE_STRUCTURE } from "src/utils/constants";

const customRender = (ui, options = {}) => {
  const { routerInitialIndex = 0 } = options;

  const AllTheProviders = ({ children }) => {
    const routes = [
      {
        path: "/",
        element: children,
        id: PAGE_STRUCTURE.home.clientPageId,
        loader: () => ({
          productList,
          cart,
        }),
        handle: {
          crumb: () => ({ url: "/", content: "home route" }),
        },
        children: [
          {
            path: "/:productId",
            element: children,
            id: PAGE_STRUCTURE.product.clientPageId,
            loader: () => productInfo,
            handle: {
              crumb: () => ({ url: "/", content: "product route" }),
            },
          },
        ],
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/", "/123"],
      initialIndex: routerInitialIndex,
    });
    return (
      <RouterProvider router={router}>
        <ProductContextProvider
          value={{
            list: productList,
            cart: cart,
          }}
        >
          {children}
        </ProductContextProvider>
      </RouterProvider>
    );
  };

  AllTheProviders.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
export { customRender as render };
