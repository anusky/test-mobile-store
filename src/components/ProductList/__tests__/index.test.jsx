import { render, screen, waitFor } from "src/test-config/test-utils";
import ProductList from "..";
import productList from "src/test-config/__fixtures__/productList.json";
import { act } from "react-dom/test-utils";

describe("ProductList coverage", () => {
  it("ProductList renders properly", async () => {
    render(<ProductList list={productList} />);
    expect(
      await screen.findByTestId("productlist-component")
    ).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    let container;
    await act(() => {
      container = render(<ProductList list={productList} />);
    });
    expect(container.asFragment()).toMatchSnapshot();
  });

  it("ProductList should render 3 items from fixtures", async () => {
    render(<ProductList list={productList} />);

    await waitFor(() => {
      const totalLinks = screen.getAllByRole("link").length;
      expect(totalLinks).toBe(3);
    });
  });

  it("ProductList product link should contain `brand, model` format", async () => {
    // Check made with first fixture product element
    render(<ProductList list={productList} />);
    expect(
      await screen.findByRole("link", { name: /Acer, Iconia Talk S/i })
    ).toBeInTheDocument();
  });

  it("ProductList product link should point to '/:id' page", async () => {
    // Check made with first fixture product element
    render(<ProductList list={productList} />);

    await waitFor(() => {
      const links = screen.getAllByRole("link")[0];
      expect(links).toHaveAttribute("href", "/ZmGrkLRPXOTpxsU4jjAcv");
    });
  });
});
