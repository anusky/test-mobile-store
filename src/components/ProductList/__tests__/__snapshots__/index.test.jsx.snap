// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList coverage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-y-8 justify-center"
    data-testid="productlist-component"
  >
    <a
      class="grid gap-y-4 gap-x-4 justify-center hover:bg-slate-50 rounded-lg py-4 px-3"
      href="/ZmGrkLRPXOTpxsU4jjAcv"
    >
      <div
        class="object-cover"
      >
        <img
          alt="Acer, Iconia Talk S"
          src="https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
        />
      </div>
      <div
        class="grid"
      >
        <p>
          Acer, Iconia Talk S
        </p>
        <p
          class="font-bold"
        >
          170€
        </p>
      </div>
    </a>
    <a
      class="grid gap-y-4 gap-x-4 justify-center hover:bg-slate-50 rounded-lg py-4 px-3"
      href="/cGjFJlmqNPIwU59AOcY8H"
    >
      <div
        class="object-cover"
      >
        <img
          alt="Acer, Liquid Z6 Plus"
          src="https://front-test-api.herokuapp.com/images/cGjFJlmqNPIwU59AOcY8H.jpg"
        />
      </div>
      <div
        class="grid"
      >
        <p>
          Acer, Liquid Z6 Plus
        </p>
        <p
          class="font-bold"
        >
          250€
        </p>
      </div>
    </a>
    <a
      class="grid gap-y-4 gap-x-4 justify-center hover:bg-slate-50 rounded-lg py-4 px-3"
      href="/8hKbH2UHPM_944nRHYN1n"
    >
      <div
        class="object-cover"
      >
        <img
          alt="Acer, Liquid Z6"
          src="https://front-test-api.herokuapp.com/images/8hKbH2UHPM_944nRHYN1n.jpg"
        />
      </div>
      <div
        class="grid"
      >
        <p>
          Acer, Liquid Z6
        </p>
        <p
          class="font-bold"
        >
          120€
        </p>
      </div>
    </a>
    <div />
  </div>
</DocumentFragment>
`;
